src/frequenz/client/reporting/_client.py

```diff
@@ -0,0 +1,289 @@
+)
+"""Type for a sample of a time series incl. metric type, microgrid and component ID"""
+
+
+@dataclass(frozen=True)
+class ComponentsDataPage:
+    """A page of microgrid components data returned by the Reporting service."""
+
+    _data_pb: PBListMicrogridComponentsDataResponse
+    """The underlying protobuf message."""
+
+    def is_empty(self) -> bool:
+        """Check if the page contains valid data.
+
+        Returns:
+            True if the page contains no valid data.
+        """
+        if not self._data_pb.microgrids:
+            return True
+        if not self._data_pb.microgrids[0].components:
+            return True
+        if not self._data_pb.microgrids[0].components[0].metric_samples:
+            return True
+        return False
+
+    def iterate_metric_samples(self) -> Generator[MetricSample, None, None]:
+        """Get generator that iterates over all values in the page.
+
+        Note: So far only `SimpleMetricSample` in the `MetricSampleVariant`
+        message is supported.
+
+
+        Yields:
+            A named tuple with the following fields:
+            * timestamp: The timestamp of the metric sample.
+            * microgrid_id: The microgrid ID.
+            * component_id: The component ID.
+            * metric: The metric name.
+            * value: The metric value.
+        """
+        data = self._data_pb
+        for mdata in data.microgrids:
+            mid = mdata.microgrid_id
+            for cdata in mdata.components:
+                cid = cdata.component_id
+                for msample in cdata.metric_samples:
+                    ts = msample.sampled_at.ToDatetime()
+                    met = Metric.from_proto(msample.metric).name
+                    value = (
+                        msample.sample.simple_metric.value
+                        if msample.sample.simple_metric
+                        else None
+                    )
+                    yield MetricSample(
+                        timestamp=ts,
+                        microgrid_id=mid,
+                        component_id=cid,
+                        metric=met,
+                        value=value,
+                    )
+
+    @property
+    def next_page_token(self) -> Any:
+        """Get the token for the next page of data.
+
+        Returns:
+            The token for the next page of data.
+        """
+        return self._data_pb.pagination_info.next_page_token
+
+
+class ReportingClient:
+    """A client for the Reporting service."""
+
+    def __init__(self, service_address: str):
+        """Create a new Reporting client.
+
+        Args:
+            service_address: The address of the Reporting service.
+        """
+        self._grpc_channel = grpcaio.insecure_channel(service_address)
+        self._stub = ReportingStub(self._grpc_channel)
+
+    # pylint: disable=too-many-arguments
+    async def iterate_single_metric(
```

Comment:
I wouldn't name these methods `iterate`, not sure if they are a common idiom in data science but otherwise I'd say it is not a common idiom in general in python. I think client methods should map more or less one to one to the gRPC API methods. In this case I think the name should be: `list_microgrid_components_data()`. I think it is fine to do the page unwrapping and still use the same name as the gRPC API method.

Again, I think we should have a common approach in all APIs.

PS: I guess the extra explicitness in names might come from *untyped python*, in that case it makes more sense, but when using type hints you can already see you are iterating (`AsyncIterator`) over a single metric (`[Sample]`).

Reply:
IMHO it is pretty clear that Python API client method names should match the gRPC API, so I suggest to create a separate issue about this only if anyway thinks differently.

As for how to iterate, please refer to:

* https://github.com/frequenz-floss/frequenz-client-base-python/discussions/41

Reply:
>I wouldn't name these methods iterate,

> PS: I guess the extra explicitness in names might come from untyped python, in that case it makes more sense, but when using type hints you can already see you are iterating (AsyncIterator) over a single metric ([Sample]).

Agree, can remove this. Actually the explicitness is to distinguish from different access methods that I considered, e.g. alternative to the iterator a dictionary representing all nesting levels. At the moment I am not sure whether it will come.

> In this case I think the name should be: list_microgrid_components_data()

Not sure about this. The method offers only a small subset of the potential functionality of the gRPC function. And the  current name reflects this limitation. Also, since the current version is sufficient for many applications already, I am not sure whether this method will be extended or another method will be added, that provides full support of the gRPC.

Reply:
Mmmm, OK, I see. In general I think clients should only be a thin wrapper over the gRPC calls, and other more convenient functions could be built on top. I mean, I'm completely fine to keep it as it is for an early version, especially if we need something fast, but in the long term I don't think it is the right approach. If the API is more flexible, then some users must need that flexibility, otherwise the gRPC API we have is more complex than needed.

Reply:
>If the API is more flexible, then some users must need that flexibility, otherwise the gRPC API we have is more complex than needed.

Yes, the flexibility of the API will be exposed, there are issues for all of the missing features. Question is whether we have only one method that exposes all or multiple. I would postpone the decision until we address some of the open issues, e.g. https://github.com/frequenz-floss/frequenz-client-reporting-python/issues/19.

Reply:
Partly addressed here: https://github.com/frequenz-floss/frequenz-client-reporting-python/pull/33.

On whether to keep the existing single component function, see https://github.com/frequenz-floss/frequenz-client-reporting-python/issues/34.

On the iterator, see https://github.com/frequenz-floss/frequenz-client-base-python/discussions/41.